Feature/gRPC

the-man-w-laughs opened the pull request:
Change http to gRPC in gathering FullProjectInfo form TeamHub for ReportHub.

Tech-stack:
- code first approach to generate gRPC service and client;

the-man-w-laughs committed:
Add initial configuration of gRPC to TeamHub.

the-man-w-laughs committed:
Add gRPC client to ReportHub.

mebr1 commented on src/Shared/Extensions/ClaimsPrincipalExtensions.cs:
```diff
@@ -24,6 +25,7 @@ public static int GetUserId(this ClaimsPrincipal principal)
+
         if (claim == null)
         {
             throw new InvalidOperationException("The 'sub' claim is not present in the principal.");
```
It's not necessary at the moment, but it's better to move exception messages to a separate file

the-man-w-laughs replied:
What if I need to throw an exceptions with interpolated string messages or create a class only for parameterless messages?

mebr1 replied:
You can create custom exceptions and overload their constructors with various parameters. Within the service, you simply call the required exception constructor, eliminating the need to explicitly write the message. For constant messages, you can create a separate class and directly insert them into the exception

mebr1 commented on src/services/ReportHub/ReportHub.BLL/Extensions/FullProjectParserExtensions.cs:
```diff
@@ -1,23 +1,26 @@
-        public static string ToReport(this FullProjectResponseDto project)
+        public static string ToReport(this FullProjectInfoResponse project)
         {
             var sb = new StringBuilder();
```
Let's make a more specific variable name

mebr1 replied:
I mean that the variable name "sb" isn't informative. While you work with it extensively, it's difficult to determine its contents from the name alone

mebr1 commented on src/services/ReportHub/ReportHub.BLL/Services/ProjectReportService.cs:
```diff
@@ -50,16 +51,18 @@ public async Task<FileStreamResult> GetLatestProjectReportAsync(int projectId)
             //     return latestReport;
```
Remove the commented code

mebr1 commented on src/services/ReportHub/ReportHub.BLL/Services/ProjectReportService.cs:
```diff
@@ -50,16 +51,18 @@ public async Task<FileStreamResult> GetLatestProjectReportAsync(int projectId)
-            );
+            var request = new FullProjectInfoRequest() { ProjectId = projectId, UserId = userId };
+
+            FullProjectInfoResponse fullProjectResponseDto =
```
Use var, don't explicitly specify the type

mebr1 commented on src/services/ReportHub/ReportHub.BLL/Services/ProjectReportService.cs:
```diff
@@ -50,16 +51,18 @@ public async Task<FileStreamResult> GetLatestProjectReportAsync(int projectId)
                 );
             }
+
             string reportContent = fullProjectResponseDto.ToReport();
```
Same. Check throughout the code for explicit type annotations

mebr1 commented on src/services/TeamHub/TeamHub.BLL/Extensions/GrpcRegistrationExtensions.cs:
```diff
@@ -0,0 +1,52 @@
+
+        public static WebApplicationBuilder ConfigureWebHost(this WebApplicationBuilder builder)
+        {
+            builder.WebHost.ConfigureKestrel(options =>
```
Move the Kestrel configuration to the appsettings file

the-man-w-laughs replied:
Cannot find any info how to do that.

mebr1 replied:
https://learn.microsoft.com/en-us/aspnet/core/fundamentals/servers/kestrel/endpoints?view=aspnetcore-7.0

mebr1 commented on src/services/TeamHub/TeamHub.DAL/Extensions/DalRegistrationExtensions.cs:
```diff
@@ -17,8 +17,10 @@ IConfiguration configuration
             {
                 var connectionString = configuration.GetConnectionString("Default");
+
                 if (string.IsNullOrWhiteSpace(connectionString))
```
Add curly braces for if

mebr1 commented on src/services/ReportHub/ReportHub.BLL/Extensions/GrpcRegistrationExtensions.cs:
```diff
@@ -0,0 +1,30 @@
+            this IServiceCollection services,
+            IConfiguration config
+        )
+        {
```
Let's use AddGrpcClient for configuration

the-man-w-laughs replied:
I use code first approach to configure gRPC and no client code is generated.
I didn't find any overloads or info how to inject client using AddGrpcClient if i don't have a client class.

mebr1 commented on src/services/ReportHub/ReportHub.BLL/Extensions/Utils/GrpcClientFactory.cs:
```diff
@@ -0,0 +1,21 @@
+
+namespace ReportHub.BLL.Extensions.Utils
+{
+    public class GrpcClientFactory
```
Where are you using this class? I didn't find any references to it

the-man-w-laughs replied:
I removed it

mebr1 commented on src/services/TeamHub/TeamHub.BLL/Extensions/GrpcRegistrationExtensions.cs:
```diff
@@ -0,0 +1,52 @@
+
+        public static void UseGrpcService(this IApplicationBuilder app)
+        {
+            app.UseRouting();
```
UseRouting should not be placed here

the-man-w-laughs committed:
Refactor code

Major changes:
- remove explicit typing;
- remove unused GrpcClientFactory;
- remove commented code;
- move UseRouting to program.cs;

the-man-w-laughs committed:
Refactor code.

Major changes:
- rename Stringuilder from sb to report;
- move kestrel configuration from code to appsettings;

the-man-w-laughs committed:
Merge branch 'feature/gateway' into feature/gRPC
